### ISSUE TITLE
Restarting a WebJob that writes to Blob Storage in a Function can lead to data loss.

### ISSUE THREAD

**Issue by alxbl** [Azure/azure-storage-net#539]
I feel like I'm most likely missing some obvious setting or doing something very subtly wrong, 
but haven't been able to figure this out in the past week or so, so here goes:

> I'm not sure if this should be filed as a WebJobs, ServiceBus, or Storage bug.
> Let me know if I should open it elsewhere.

# Restarting a WebJob that writes to Blob Storage in a `Function` can lead to data loss.

It looks like when a WebJob that writes to blob storage is restarted, there is a race condition (?)
where topic messages can be marked as completed (We use `AutoComplete = true`) but the blob
isn't actually present in the storage.

This seems to happen as soon as the WebJob needs to restart, so all the following lead to repro:

- Application Settings Changes (On Save)
- Scale Up (e.g. S1 -> S3)
- Scale Out (e.g. 1 instance -> 10 instances)
- Scale In (e.g. 10 instances -> 1 instance)
- Scale Down (e.g. S3 -> S1)
- Manual WebJob Restart (Restart and/or Stop then Start)

This doesn't seem to happen when restarting the WebApp that is hosting the WebJob. I believe this is 
normal because the WebJob is running on the Service Plan and not on the WebApp itself 
(Correct me if I'm wrong.)

I wrote [a sample application][2] that does away with all of our business logic and narrows it down to one of
the Azure SDKs (or my ability to C#) for code review or if someone would like to try to reproduce the 
issue on their end.

[2]: https://github.com/alxbl/ServiceBus.MessageLoss

The best way to reproduce this issue is to play around with the application settings, scale and to 
manually restart the webjob while it is processing messages.

> For an average test run, I see roughly 2000-5000 blobs out of ~100K that would simply not be in 
> blob storage, but also not be in any dead letter queue or elsewhere. This varies depending on
> how many times I trigger web job restarts as mentioned above.

> Reproducing can be a little time consuming, so what I usually do is setup an S3 with 10 instances of the webjob
> and change the application settings/restart a few times while I watch the topic queue length go down in
> ServiceBusExplorer. When the message count is low enough, I'll scale down to S1 1 instance until the queues
> hit 0 and stop the job before running the validation script (see pseudocode below)

## Configuration

**Storage Account**

- Performance = `Standard`
- Secure Transfer = `Disabled`
- Replication = `LRS`

**WebJob**

- Topic:
    - Enabled Batched Messages = `true`
    - Enabled Partitioning = `true`
- Subscription:
    - Enabled Batched Messages = `true`
    - Deadletter on filter evaluation errors = `true`
- Max Concurrent Calls = `32`
- Prefetch Count = `64`
- Auto-Renew Timeout = `120 seconds`
- Always On (Web App) = `On`

## Sample App Pseudo-Code

**WebJob**

1. Call the destination container `numbers`.
1. Let there be messages `1, 2, ..., N` in the topic `topic/sub`. `N` is the expected number of blobs.
1. For message `m` in `topic/sub`,
    1. Create a blob named `m` with arbitrary content in the `numbers` container. 
       (`await b.UploadTextAsync(m.ToString(), 'true')`)
    1. Mark message `m` as complete (`<auto-complete>`)
 

**Validation**

1. Retrieve all the blobs in `numbers`, call the collection `blobs`.
1. Assert  `blobs.Count` == `N`.
1. Let `expected` be the collection of numbers `1..N`.
1. Assert `expected.Except(blobs).Count == 0`.
1. If any, print `expected.Except(blobs)`.

(Assertions fail and a list of blobs is printed.)

## What we've tried so far

- Stripped away all of the business logic to rule out the application.
- Made sure cancellation tokens are passed all the way down to `CloudBlockContainer.UploadTextAsync()`.
- Tried switching `async` calls to their `sync` version.
- Upgraded to Storage SDK 8.4.0.
- Upgraded to latest WebJobs and ServiceBus SDKs.

## What we're planning to try in parallel with this issue

- Manual message completion to try and determine if `AutoComplete` is the issue
- Use REST API directly via `HttpClient` to try and rule out the Storage SDK.
- Hacky code that attempts to verify that the blob is written before returning out of the `Function` (I would rather not resort to this in our production deployments)

## Impact

As you would guess, it's a bit unnerving to deploy to production or even enable autoscale / change settings in a production environment when it carries a risk of losing production data in the process.

## Workaround

The only workaround at this point is to stop the services that feed the topics and wait for the data to flush
before making the necessary changes, which is clumsy, not easy to automate, and not very continuous deployment friendly.

-----

Any advice or insight into something we could be doing wrong (Azure configuration, Code, or otherwise) will be very welcome.

**Comment by alxbl** [comment]
I've spent some more time on this, and I have narrowed down the issue to be with ServiceBus' AutoComplete flag.

After setting `AutoComplete = false` and handling message completion manually. we have not lost a single message out of a full (~5.5m message) run with aggressive restarting.

This makes me think that either the AutoComplete documentation is [wrong][1] or there is a bug in the WebJobs.ServiceBus library:

> Gets or sets a value that indicates whether the message-pump should call `CompleteAsync(String)` or `CompleteAsync(String)` on messages after the callback has **completed processing**.

> `true` to complete the message processing automatically on **successful execution of the operation**; otherwise, false.

(Notice that one sentence reads "successful execution", the other simply says "completed processing")

[1]: https://docs.microsoft.com/en-us/dotnet/api/microsoft.azure.servicebus.messagehandleroptions.autocomplete?view=azure-dotnet

To summarize, this is clearly not a Storage SDK bug (Sorry!), but might be a bug/documentation typo in the WebJobs.ServiceBus SDK.

We have our workaround: Don't use `AutoComplete=true` in ServiceBus triggered webjobs.

This can be closed. However, should I open a separate issue with the WebJobs SDK project?

**Comment by erezvani1529** [comment]
Hi @alxbl ,

Would you please refer to the issue you have opened with WebJobs SDK here for other interested users?

Thanks!

**Comment by alxbl** [comment]
Hi @erezvani1529, I didn't open the issue with Webjobs SDK as I was waiting for feedback on this issue. I'll open it and link to this. 

Cheers,
Alex
